### PR TITLE
Read extended BPMN model

### PR DESCRIPTION
**Pull request by Beacze**

### CODE REVIEW

File: Kitodo-DataManagement/src/main/java/org/kitodo/data/database/beans/Workflow.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+@Entity
+@Table(name = "workflow")
+public class Workflow extends BaseBean {
+    private static final long serialVersionUID = 6831844584235763486L;
+
+    @Column(name = "title")
+    private String title;
+
+    @Column(name = "file")
+    private String file;
```
[the remark this answers is not in the document]
Comment (henning-gerhardt) [reply Kitodo-DataManagement/src/main/java/org/kitodo/data/database/beans/Workflow.java]: Why not call it `fileName` every where and add maybe a explanation that this field only contains the name of file and path is taken from `kitodo_config.properties`?

File: Kitodo/src/main/java/org/kitodo/workflow/model/Reader.java [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+
+        return task;
+    }
+
+    /**
+     * Get workflow.
+     *
+     */
+    public Diagram getWorkflow() {
+        Process process = modelInstance.getModelElementsByType(Process.class).iterator().next();
```
[the remark this answers is not in the document]
Comment (Beacze) [reply Kitodo/src/main/java/org/kitodo/workflow/model/Reader.java]: Method getModelElementsByType initalize array:
```java
List<ModelElementInstance> instances = new ArrayList();
```
I'm not sure about this iterator. I suspect that here this iterator().next() can return null, but iterator alone will not cause NPE. More lilkely process passed to Diagram constructor can be null, so meybe check should be done there. Probably good idea would be to test it for such a case, so I will add tests for non existing process definition, even if such a situation shouldn't happend as it reads from list of existings defintions (stored in database).

File: Kitodo/src/main/java/org/kitodo/workflow/model/Reader.java [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+        return new Diagram(process);
+    }
+
+    private void getWorkflowTasks() {
+        StartEvent startEvent = modelInstance.getModelElementsByType(StartEvent.class).iterator().next();
```
[the remark this answers is not in the document]
Comment (Beacze) [reply Kitodo/src/main/java/org/kitodo/workflow/model/Reader.java]: Answer as above.

File: Kitodo/src/main/java/org/kitodo/workflow/model/Reader.java [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+     * @param sequenceFlowIterator
+     *            iterator containing tasks following given gateway
+     * @param ordering
+     *            which is going to be assigned to tasks in this branch
+     */
+    private void addTasksBranch(Gateway gateway, Iterator<FlowNode> sequenceFlowIterator, int ordering) {
+        int j = ordering;
+        ConditionExpression conditionExpression = null;
+
+        while (sequenceFlowIterator.hasNext()) {
```
[the remark this answers is not in the document]
Comment (Beacze) [reply Kitodo/src/main/java/org/kitodo/workflow/model/Reader.java]: sequenceFlowIterator can't be null as this.followingFlowNodes is initalized:
```java
this.followingFlowNodes = new ArrayList<>();
Iterator<FlowNode> sequenceFlowIterator = this.followingFlowNodes.iterator();
```
So here hasNext will be false and while loop will not be executed at all.

File: Kitodo/src/test/java/org/kitodo/workflow/ReaderIT.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
+package org.kitodo.workflow;
```
[the remark this answers is not in the document]
Comment (Beacze) [reply Kitodo/src/test/java/org/kitodo/workflow/ReaderIT.java]: Header added.

File: Kitodo/src/test/java/org/kitodo/workflow/ReaderTest.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
+package org.kitodo.workflow;
```
[the remark this answers is not in the document]
Comment (Beacze) [reply Kitodo/src/test/java/org/kitodo/workflow/ReaderTest.java]: Header added

File: Kitodo/src/main/java/org/kitodo/workflow/model/beans/Diagram.java [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+import org.camunda.bpm.model.bpmn.instance.Process;
+
+public class Diagram {
+
+    private String id;
+    private String title;
+    private String outputName;
+    private Integer docket;
+    private Integer ruleset;
+    private static final String NAMESPACE = "http://www.kitodo.com/template";
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/org/kitodo/workflow/model/beans/Diagram.java]: - Why are you using URI `kitodo.com` as the project is using `kitodo.org` domain?
- You are dublicating this constant value in class `KitodoTask` - should this constant with same value not defined on one place and used from both classes?

File: Kitodo/src/main/java/org/kitodo/workflow/model/beans/KitodoTask.java [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+    private Boolean typeImportFileUpload;
+    private Boolean typeExportDms;
+    private Boolean typeExportRussian;
+    private Boolean typeImagesRead;
+    private Boolean typeImagesWrite;
+    private Boolean typeAcceptClose;
+    private Boolean typeCloseVerify;
+    private Boolean batchStep;
+
+    static final String NAMESPACE = "http://www.kitodo.com/template";
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/org/kitodo/workflow/model/beans/KitodoTask.java]: See my comment on `Diagram` class to same named constant.

File: Kitodo/src/main/java/org/kitodo/workflow/model/beans/Diagram.java (hunk 7 above)
Comment (Beacze) [reply Kitodo/src/main/java/org/kitodo/workflow/model/beans/Diagram.java]: Diagram doesn't extend KitodoTask. Do you think that i would be also better to store in kitodo_config.properties?

Comment (henning-gerhardt) [reply Kitodo/src/main/java/org/kitodo/workflow/model/beans/Diagram.java]: - If this value should be changed by user / administrator then it belongs to kitodo_config.properties.
- If this value should not be changed from outside, then it should stay here.

So, at least change the URI value to `kitodo.org`.

Comment (Beacze) [reply Kitodo/src/main/java/org/kitodo/workflow/model/beans/Diagram.java]: No, basically it shouldn't be changed, as it is defined in modeler.

**Pull request merged**